src/Aspire.Hosting/ApplicationModel/CommandsConfigurationExtensions.cs

```diff
@@ -112,9 +135,59 @@ internal static void AddLifeCycleCommands(this IResource resource)
             iconVariant: IconVariant.Regular,
             isHighlighted: false));
 
-        static bool IsStopped(string? state) => state is "Exited" or "Finished" or "FailedToStart";
+        static bool IsStopped(string? state) => state is "Exited" or "Finished";
+        static bool IsStartable(string? state) => IsStopped(state) || state is "FailedToStart";
         static bool IsStopping(string? state) => state is "Stopping";
         static bool IsStarting(string? state) => state is "Starting";
+        static bool IsRunning(string? state) => state is "Running";
         static bool IsWaiting(string? state) => state is "Waiting" or "RuntimeUnhealthy";
     }
+
+    internal static async Task<ExecuteCommandResult> RunLifeCycleCommandAsync(
+        Func<IServiceProvider, CancellationToken, Task> action,
+        IServiceProvider serviceProvider,
+        string resourceName,
+        Func<string, bool> isExpectedStateFunc,
+        TimeSpan timeout,
+        CancellationToken cancellationToken)
+    {
+        using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+        cts.CancelAfter(timeout);
+
+        var resourceNotificationService = serviceProvider.GetRequiredService<ResourceNotificationService>();
+
+        try
+        {
+            await action(serviceProvider, cts.Token).ConfigureAwait(false);
+
+            await WaitForResourceAsync(resourceNotificationService, resourceName, isExpectedStateFunc, cts.Token).ConfigureAwait(false);
```

Comment:
What are the implications of long running blocking commands to the UI and grpc clals? Can I spam and overlap these calls just fine?

Reply:
This is in a gRPC call and yes it's fine.